velox/exec/AssignUniqueId.cpp

```diff
@@ -0,0 +1,109 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/AssignUniqueId.h"
+
+#include <utility>
+
+namespace facebook::velox::exec {
+
+AssignUniqueId::AssignUniqueId(
+    int32_t operatorId,
+    DriverCtx* driverCtx,
+    const std::shared_ptr<const core::AssignUniqueIdNode>& planNode,
+    int32_t uniqueTaskId,
+    std::shared_ptr<std::atomic_int64_t> rowIdPool)
+    : Operator(
+          driverCtx,
+          planNode->outputType(),
+          operatorId,
+          planNode->id(),
+          "AssignUniqueId") {
+  isIdentityProjection_ = true;
+  rowIdPool_ = std::move(rowIdPool);
+
+  uniqueValueMask_ = ((int64_t)uniqueTaskId) << 40;
+  rowIdCounter_ = rowIdPool_->fetch_add(rowIdsPerRequest_);
+  maxRowIdCounterValue_ =
+      std::min(rowIdCounter_ + rowIdsPerRequest_, maxRowId_);
+}
+
+void AssignUniqueId::addInput(RowVectorPtr input) {
+  auto numInput = input->size();
+  VELOX_CHECK_NE(
+      numInput, 0, "AssignUniqueId::addInput received empty set of rows")
+  input_ = std::move(input);
+}
+
+RowVectorPtr AssignUniqueId::getOutput() {
+  if (input_ == nullptr) {
+    return nullptr;
+  }
+
+  BufferPtr nulls;
+
+  if (input_->nulls() != nullptr) {
+    nulls =
+        AlignedBuffer::allocate<bool>(input_->size() + 1, operatorCtx_->pool());
+    auto inputNulls = input_->nulls()->as<uint64_t>();
+    auto outputNulls = nulls->asMutable<int64_t>();
+
+    for (int i = 0; i < input_->size(); i++) {
+      bits::setBit(outputNulls, i, bits::isBitSet(inputNulls, i));
+    }
+    bits::setBit(outputNulls, input_->size(), false);
+  }
+
+  auto ids = generateIdColumn();
+  std::vector<VectorPtr> children;
+  for (const auto& child : input_->children()) {
+    children.emplace_back(child);
+  }
+  children.emplace_back(ids);
+
+  input_ = nullptr;
+  return std::make_shared<RowVector>(
+      operatorCtx_->pool(), outputType_, nulls, children[0]->size(), children);
+}
+
+void AssignUniqueId::finish() {
+  if (!isFinishing_ && input_ == nullptr) {
+    Operator::finish();
+  }
+}
+
+FlatVectorPtr<int64_t> AssignUniqueId::generateIdColumn() {
+  size_t inputSize = input_->children()[0]->size();
+  BufferPtr values =
+      AlignedBuffer::allocate<int64_t>(inputSize, operatorCtx_->pool());
+  auto* valuesPtr = values->asMutable<int64_t>();
+
+  for (int i = 0; i < inputSize; i++) {
+    if (rowIdCounter_ >= maxRowIdCounterValue_) {
+      rowIdCounter_ = rowIdPool_->fetch_add(rowIdsPerRequest_);
+      maxRowIdCounterValue_ =
+          std::min(rowIdCounter_ + rowIdsPerRequest_, maxRowId_);
+    }
+    auto rowId = rowIdCounter_++;
+    valuesPtr[i] = uniqueValueMask_ | rowId;
+  }
+  return std::make_shared<FlatVector<int64_t>>(
```

Comment:
consider using results_ from the base class and re-use the vector if it is singly-referenced